File: data/Stack.ts
export default class Stack<T> {
  private items: T[] = [];
  private max?: number;

  constructor(max?: number) {
    this.max = max;
  }

  public push(item: T) {
    if (this?.max !== undefined && this.items.length >= this.max) {
      return false;
    }

    this.items.unshift(item);
    return true;
  }

  public pop() {
    return this.items.shift() ?? null;
  }

  public peek() {
    const slice = this.items.slice(0, 1);

    if (slice.length === 0) {
      return null;
    }
    return slice[0];
  }

  public size() {
    return this.items.length;
  }

  [Symbol.iterator]() {
    let index = 0;

    return {
      next: () => {
        if (index < this.items.length) {
          const res = { value: this.items[index], done: false };
          index = index + 1;
          return res;
        } else {
          return { done: true };
        }
      },
    };
  }
}

File: data/stack.go
package data

type Stack struct {
	items []interface{}
	max   int
}

func NewStack(max int) *Stack {
	return &Stack{items: make([]interface{}, 0, max), max: max}
}

func (s *Stack) Push(item interface{}) bool {
	if s.max != 0 && len(s.items) >= s.max {
		return false
	}
	s.items = append([]interface{}{item}, s.items...)
	return true
}

func (s *Stack) Pop() interface{} {
	if len(s.items) == 0 {
		return nil
	}

	item := s.items[0]
	s.items = s.items[1:]
	return item
}

func (s *Stack) Peek() interface{} {
	if len(s.items) == 0 {
		return nil
	}
	return s.items[0]
}

func (s *Stack) Size() int {
	return len(s.items)
}
